plugins/node/opentelemetry-instrumentation-pg/src/pg.ts

```diff
@@ -0,0 +1,302 @@
+  }
+
+  private _getClientQueryPatch() {
+    const plugin = this;
+    return (original: typeof pgTypes.Client.prototype.query) => {
+      plugin._logger.debug(
+        `Patching ${PostgresInstrumentation.COMPONENT}.Client.prototype.query`
+      );
+      return function query(this: PgClientExtended, ...args: unknown[]) {
+        let span: Span;
+
+        // Handle different client.query(...) signatures
+        if (typeof args[0] === 'string') {
+          const query = args[0];
+          if (args.length > 1 && args[1] instanceof Array) {
+            const params = args[1];
+            span = utils.handleParameterizedQuery.call(
+              this,
+              plugin.tracer,
+              plugin._config as InstrumentationConfig &
+                PostgresInstrumentationConfig,
+              query,
+              params
+            );
+          } else {
+            span = utils.handleTextQuery.call(this, plugin.tracer, query);
+          }
+        } else if (typeof args[0] === 'object') {
+          const queryConfig = args[0] as NormalizedQueryConfig;
+          span = utils.handleConfigQuery.call(
+            this,
+            plugin.tracer,
+            plugin._config as InstrumentationConfig &
+              PostgresInstrumentationConfig,
+            queryConfig
+          );
+        } else {
+          return utils.handleInvalidQuery.call(
+            this,
+            plugin.tracer,
+            original,
+            ...args
+          );
+        }
+
+        // Bind callback to parent span
+        if (args.length > 0) {
+          const parentSpan = getSpan(context.active());
+          if (typeof args[args.length - 1] === 'function') {
+            // Patch ParameterQuery callback
+            args[args.length - 1] = utils.patchCallback(
+              span,
+              args[args.length - 1] as PostgresCallback
+            );
+            // If a parent span exists, bind the callback
+            if (parentSpan) {
+              args[args.length - 1] = context.bind(args[args.length - 1]);
+            }
+          } else if (
+            typeof (args[0] as NormalizedQueryConfig).callback === 'function'
+          ) {
+            // Patch ConfigQuery callback
+            let callback = utils.patchCallback(
+              span,
+              (args[0] as NormalizedQueryConfig).callback!
+            );
+            // If a parent span existed, bind the callback
+            if (parentSpan) {
+              callback = context.bind(callback);
+            }
+
+            // Copy the callback instead of writing to args.callback so that we don't modify user's
+            // original callback reference
+            args[0] = { ...(args[0] as object), callback };
+          }
+        }
+
+        // Perform the original query
+        const result: unknown = original.apply(this, args as never);
+
+        // Bind promise to parent span and end the span
+        if (result instanceof Promise) {
```

Comment:
i think its worth to make an utils to handle promise since both of patches handle it the same way

Reply:
I don't want to make any improvements refactoring etc. during moving to instrumentations. Don't want to break anything or make it worse then it is currently. We are just changing minimum for easier reviewing and to keep consistency as I'm also not the author of original plugin I don't want to break something. We can handle this later whoever will be eager to improve it.